# miRNA Profiling Pipeline

The TCGA miRNAseq data generation process, including strand-specific library construction, sequencing, and computational processing is described in:
Chu A, Robertson G, Brooks D, Mungall AJ, Birol I, Coope R, Ma Y, Jones S, Marra MA. Large-scale profiling of microRNAs for The Cancer Genome Atlas. Nucleic Acids Res. 2015 Aug 13.
https://doi.org/10.1093/nar/gkv808

## Pipeline Overview

The analysis pipeline takes as input a .sam file containing BWA alignments of trimmed reads for a sample and generates a) annotations of the reads, b) expression profiles of features annotated in the sample, and c) miRNA by sample expression matrices, and d) graphs of overall feature representation, and quality metrics.

This pipeline was initially written in Perl and R, but has been re-written in Python in release 1.0. The Python package is called mirna_profiling. Analysis configurations are stored in mirna_profiling/configuration/annotation_config.yaml, which can be used as a template for customized analysis. Under workflow/ directory, there is an example workflow written in Nextflow for batch processing. The workflow runs a containerized version of the package.

The pipeline release 1.0 has been tested with BWA 0.5.7, Python 3.8 and Nextflow 21.05.0 on Centos7 machines.

## Pre-alignment processing

Before any alignment can occur, some pre-processing of the sequences is necessary. MiRNAs tend to be shorter than the length of reads generated by sequencers, which means that the read will sequence past biological RNA into the 3' adapter of the sequencing protocol. This non-biological sequence at the end of the read will make it difficult for the aligner to align the read, so it is trimmed off first.

Documentation and code to trim adapters from miRNA reads and generate adapter report is available at: http://www.bcgsc.ca/platform/bioinfo/software/adapter-trimming-for-small-rna-sequencing

The adapter report summarizes the adapter trimming done on the reads before alignment. It is a space-delimited 2-column file, containing read lengths and number of reads, and sorted by read length in ascending order.

If you are working with bam files where the reads have already been trimmed and aligned, you can generate this summary report directly from the bam with the following command:
```bash
samtools view <abc.bam> | awk '{arr[length($10)]+=1} END {for (i in arr) {print i" "arr[i]}}' | sort -t " " -k1n
```
Information from the adapter report, if available, would be included in alignment_stats.csv generated after miRNA profiling is complete.

## Running miRNA Profiling Analysis

To run the profiling analysis on multiple samples, it is best to set up the directory structure such that the scripts can automatically find the files to process. Under the project base directory ({$PROJDIR}), place all your bam files, typically one subdirectory per sample. The bam files should be named LIBRARY.bam (or LIBRARY_INDEX.bam for multiplexed runs), optionally accompanied by a LIBRARY_adapter.report (or LIBRARY_INDEX_adapter.report).


## Pipeline Scripts

1. annotate_bam -p {$PROJDIR} -u <UCSC genome, e.g. hg38> -m <miRBase version, e.g. mirna_21> -s <miRBase species code, e.g. hsa>

Creates an annotated version of the bam/sam file named .[bs]am.annot, in which tags XC, XI, and XD are added to the reads, and contain features from UCSC and miRBase overlapping the reads.

2. alignment_stat -p {$PROJDIR} -u <UCSC genome, e.g. hg38> -m <miRBase version, e.g. mirna_21> -s <miRBase species code, e.g. hsa>

Generates expression files for miRNA and other features in the sample(s).

3. expression_matrix -p {$PROJDIR} -m <miRBase version, e.g. mirna_21> -s <miRBase species code, e.g. hsa>

Generates expression matrices that summarize the expression of each miRNA primary transcript and miRNA mature strand in the sample(s).

4. project_summary -p {$PROJDIR}

Generates graphs to provide a visual summary of samples in {$PROJDIR}.

The alignment_stat program calls annotate_bam internally and generate expression files that are required by expression_matrix and project_summary.

Output files from this package are described in DATA_DESCRIPTION.md.

## Running Pipeline in Batch Mode Using Nextflow (21.05.0 or later)

```
nextflow run /<profiling_installation_dir>/workflow/main.nf -profile cluster,container --proj_dir {$PROJDIR} --out_dir /path/output_dir

Mandatory arguments:
 -profile                   cluster,container : run container on cluster
                            container :  run container on localhost
                            cluster,local : run installed package (venv) on cluster
                            local :  run installed package (venv) on localhost

 --proj_dir                 Directory where bams are stored
 --out_dir                  Output directory for profiling results
 --ucsc                     UCSC genome or database name. Default: hg38
 --mirbase                  miRBase release or database name. Default: mirna_21
 --species                  miRBase species code.  Default: hsa

Optional arguments:
--sample_csv                CSCV file with sample, bam and adapter columns
--proj_name                 Project name to use in graphs
--config_file               Custom config yaml for profiling. Must be in the same directory as main.nf,
                            if run with container profile
```

# miRNA FAQ

## 1. What are precursors? What are mature strands?

A miRNA mature strand is the functional piece of miRNA that is transcribed from one arm of the precursor miRNA. Historically, one of the arms of the precursor express the "mature strand", and the other arm expresses a "star strand" which is thought to be nonfunctional and degraded. It has now been shown that both arms express functional miRNAs, and so both arms are now known as mature strands. On miRBase, the precursor miRNA has a MIxxxxxxx accession, whereas the mature strands have a MIMATxxxxxxx accession.


## 2. What is crossmapping?

The trimmed sequence reads largely represent isomiRs, so are short (22±3 nt). Their 5’ and 3’ ends can differ from miRBase reference mature strand coordinates, particularly at the 3’ end. As well, some miRNAs occur as families of closely related sequences that have identical or nearly identical mature strand sequences (and MIMAT accession IDs). These factors result in reads crossmapping and multimapping, which we address as follows (see the publication for more details).

A short isomiR read may map exactly to mature strands whose sequences are similar but not identical, when the read sequence does not capture the bases that distinguish these miRNAs (e.g. hsa-mir-30a at 6q13 and hsa-mir-30e at 1p34.2, which differ at position 18). We report such a read as cross-mapped, and we increment the read count for each MIMAT that it mapped to.

A read can multimap to identical mature strands from a miRNA family whose members are in different locations in the genome (e.g. miR-181a-5p=MIMAT0000256 is present in hsa-mir-181a-1 at 1q32.1 and in hsa-mir-181a-2 at 9q33.3). When we annotate a read as miR-181a-5p, we increment the read count for this MIMAT, and we increment the read count of one of the genomic locations for the family’s stem-loops. See below.


## 3. What are the different miRNAs with the same mature strand?

Many miRNAs have names in the format of hsa-mir-21, ie. organism-"mir"-microRNA. However, some miRNA names have an additional suffix, eg. hsa-mir-101-1 and hsa-mir-101-2. Note that miRNAs such as hsa-mir-29a and hsa-mir-29b are not an example of this; they are simply related, but distinct miRNAs.

These additional -1/-2 suffixes denote functionally identical miRNAs expressed from different loci. The miRBase entries for hsa-mir-101 provides the following information:
hsa-mir-101-1; accession MI0000103; chr1: 65524117-65524191 \[-\]
 Mature sequence hsa-miR-101-5p; accession MIMAT0004513
 Mature sequence hsa-miR-101-3p; accession MIMAT0000099
 hsa-mir-101-2; accession MI0000739; chr9: 4850297-4850375 \[+\]
 Mature sequence hsa-miR-101-3p; accession MIMAT0000099

The mature sequence hsa-miR-101-3p has the same accession (and sequence) in both hsa-mir-101-1 and 101-2. This is the case for all miRNAs with only differing suffixes.

During miRNA profiling, if a read sequence has enough bases outside the canonical mature sequence to unambiguously assign it to a particular miRNA ID, then it does so. However, if the read maps to the mature strand, eg. MIMAT0000099, then there is no way to unambiguously assign it as hsa-mir-101-1 or 101-2, and one of the miRNA IDs is arbitrarily (not randomly) assigned. For this reason, when working at the precursor level, it is suggested that miRNA counts are summed to get one count for hsa-mir-101. When working at any higher resolution, there is no ambiguity when working with mature strand names or accessions. Counts can be summed by mature accession, and the precursor name can be ignored.

## 4. Which read alignments are counted for miRNA expression profiles?

For a read annotation to be used the following rules are applied by examining the NM, X1, X0, XA, XC, and or XD fields in an annotated SAM file.

NM tag: number of "edits" for the main alignment to the reference

Tags with information for alternate alignments
X0 number of what BWA considered to be the "best" alignments for the read
X1 number of BWA secondary alignments
XA contains semicolon separated fields that each contain comma-separated info about each alternate read alignment, namely, chromosome, position, cigar, and NM info.

Fields added by annotation script:
XC general annotations for aligned reads, comma separated.
XD specific annotations for aligned reads, commar separated.
The first field of XC and XD is the annotation for the primary alignment, while subsequent fields are annotations for secondary alignments.

*Rules for counting a read alignment*

-	The minimum number of mismatches (NM tag) for all alignments for a read must be zero
-	The number of "best" alignments (X0 tag) must be 3 or fewer
-	The number of alignments considered is the number of alignment info fields (semicolon-separated XA tags), or the number of primary alignments (value of X0), which ever is smaller.

## 5. How are the reads annotated

- Annotations for read alignments considered are read from the XD field, or are read from the XC field if there is no XD field.
- Every read alignment (including primary and secondary) is annotated with the feature that has the highest annotation priority. Feature priority is discussed in the pipeline paper.
- Every read then is annotated using the feature with the highest priority across all the annotated read alignments.
